components/role-mgt/org.wso2.carbon.identity.role.mgt.core/src/main/java/org/wso2/carbon/identity/role/mgt/core/RoleManagementService.java

```diff
@@ -0,0 +1,168 @@
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.role.mgt.core;
+
+import java.util.List;
+
+/**
+ * OSGi service interface which use to manage roles.
+ */
+public interface RoleManagementService {
+
+    /**
+     * Add a new role.
+     *
+     * @param roleName     Role name.
+     * @param userList     List of users.
+     * @param groupList    List of groups.
+     * @param permissions  List of permissions.
+     * @param tenantDomain Tenant domain.
+     * @throws IdentityRoleManagementException IdentityRoleManagementException.
+     */
+    void addRole(String roleName, List<String> userList, List<String> groupList, List<String> permissions,
+            String tenantDomain) throws IdentityRoleManagementException;
+
+    /**
+     * Retrieve available roles.
+     *
+     * @param tenantDomain Tenant domain.
+     * @return List of role names.
+     * @throws IdentityRoleManagementException IdentityRoleManagementException.
+     */
+    List<String> getRoles(String tenantDomain) throws IdentityRoleManagementException;
+
+    /**
+     * Retrieve available roles.
+     *
+     * @param limit        Limit value.
+     * @param offset       Offset value.
+     * @param tenantDomain Tenant domain.
+     * @return List of role names.
+     * @throws IdentityRoleManagementException IdentityRoleManagementException.
+     */
+    List<String> getRoles(int limit, int offset, String tenantDomain) throws IdentityRoleManagementException;
+
+    /**
+     * Retrieve available roles.
+     *
+     * @param filter       Filter for the role name.
+     * @param limit        Limit value.
+     * @param offset       Offset value.
+     * @param sortBy       SortBy value.
+     * @param sortOrder    Sort order value.
+     * @param tenantDomain Tenant domain.
+     * @return List of role names.
+     * @throws IdentityRoleManagementException IdentityRoleManagementException.
+     */
+    List<String> getRoles(String filter, int limit, int offset, String sortBy, String sortOrder, String tenantDomain)
+            throws IdentityRoleManagementException;
+
+    /**
+     * Retrieve the given role.
+     *
+     * @param roleName     Role name.
+     * @param tenantDomain Tenant domain.
+     * @return The role object.
+     * @throws IdentityRoleManagementException IdentityRoleManagementException.
+     */
+    Role getRole(String roleName, String tenantDomain) throws IdentityRoleManagementException;
```

Comment:
since role has an 'id' shall we add a method to retrieve roles by id?

Reply:
Role ID is generated in the SCIM layer and store in SCIM tables. In this level, the role does not have an ID as we use the existing `UM_HYBRID_ROLE` table to store roles.